src/vdaf/mastic.rs

```diff
@@ -379,32 +382,477 @@ where
+            Some(self.szk.query(
+                root_share.as_ref(),
+                input_share.proof_share.clone(),
+                verify_key,
+                nonce,
+            )?)
+        } else {
+            None
+        };
+        let (prep_share, prep_state) =
+            if let Some((szk_query_share, szk_query_state)) = szk_verify_opt {
+                let verifier_len = szk_query_share.flp_verifier.len();
+                (
+                    MasticPrepareShare {
+                        vidpf_proof: xof.into_seed(),
+                        szk_query_share_opt: Some(szk_query_share),
+                    },
+                    MasticPrepareState {
+                        output_shares: MasticOutputShare::<T::Field>::from(output_shares),
+                        szk_query_state,
+                        verifier_len: Some(verifier_len),
+                    },
+                )
+            } else {
+                (
+                    MasticPrepareShare {
+                        vidpf_proof: xof.into_seed(),
+                        szk_query_share_opt: None,
+                    },
+                    MasticPrepareState {
+                        output_shares: MasticOutputShare::<T::Field>::from(output_shares),
+                        szk_query_state: None,
+                        verifier_len: None,
+                    },
+                )
+            };
+        Ok((prep_state, prep_share))
+    }
+
+    fn prepare_shares_to_prepare_message<
+        M: IntoIterator<Item = MasticPrepareShare<T::Field, SEED_SIZE>>,
+    >(
+        &self,
+        _agg_param: &MasticAggregationParam,
+        inputs: M,
+    ) -> Result<MasticPrepareMessage<T::Field, SEED_SIZE>, VdafError> {
+        let mut inputs_iter = inputs.into_iter();
+        let leader_share = inputs_iter.next().ok_or(VdafError::Uncategorized(
+            "No leader share received".to_string(),
+        ))?;
+        let helper_share = inputs_iter.next().ok_or(VdafError::Uncategorized(
+            "No helper share received".to_string(),
+        ))?;
+        if inputs_iter.next().is_some() {
+            return Err(VdafError::Uncategorized(
+                "more than 2 prepare shares".to_string(),
+            ));
+        };
+
+        match (leader_share, helper_share) {
+            (
+                MasticPrepareShare {
+                    vidpf_proof: leader_vidpf_proof,
+                    szk_query_share_opt: Some(leader_query_share),
+                },
+                MasticPrepareShare {
+                    vidpf_proof: helper_vidpf_proof,
+                    szk_query_share_opt: Some(helper_query_share),
+                },
+            ) => {
+                if leader_vidpf_proof == helper_vidpf_proof {
+                    Ok(Some(SzkQueryShare::merge_verifiers(
+                        leader_query_share,
+                        helper_query_share,
+                    )))
```

Comment:
Hmm, it looks like we pass the FLP verifier to `prep_next()` below. This is not what the spec says to do; instead we call FLP decide here and delay the joint randomness check to the next step: https://jimouris.github.io/draft-mouris-cfrg-mastic/draft-mouris-cfrg-mastic.html#section-4.2-13